### PR TITLE
Http framework free

### PR DESCRIPTION
**Pull request by trygve-lie**
This PR makes the Layout http framework free making it possible to integrate the Layout with multiple http frameworks such as Hapi, Fastify etc.

Previously Podium has been very tied to Express.js where each module in the request chain has been a middleware where properties from each module has been written to `res.locals.podium`. Internally multiple of Express.js only methods have also been used making Podium very tied to Express.js.

This implementation switches from using a middleware chain passing request and response objects around to a Promise chain where a HttpIncoming object we have controll over is passed on. In other words; each module in Podium does now have a `.process()` method which can be chained together making it possible for any other http framwork to be integrated.

All usage of Express.js only methods has also been replaced with generic methods which should work within any http framework.

Currently there is a Hapi plugin using this: https://github.com/podium-lib/hapi-layout

Despite making Podium http framework free, we treat Connect middleware based http frameworks (such as Express.js) first class by keeping the already existing `.middleware()` method. This method does now just wrap the `.process()` method. Due to this its easy to get started while it also provide backward compabillity.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -218,17 +225,57 @@ const layout = new Layout({
 
 The Layout instance has the following API:
 
-### .middleware()
+### .process(HttpIncoming)
 
-A connect compatible middleware which takes care of multiple operations needed for
-a Layout to fully work.
+Metod for processing a incomming http request. This method is intended to be
```
Comment (alexanbj) [inline README.md]: incoming

File: README.md [hunk 2]
```diff
@@ -218,17 +225,57 @@ const layout = new Layout({
...
+used to implement support for multiple http frameworks and should not really be
+used directly in a layout server.
 
 What it does:
 
--   Runs [context parsers](https://github.com/podium-lib/context) on incomming requests and creates an object on the response at `res.locals.podium.context` which can be passed on to the client when requesting content from podlets.
--   Mounts the [proxy](https://github.com/podium-lib/proxy) so that each podlet can do transparent proxy requests if needed.
+-   Runs [@podium/context] parsers on the incomming request and sets an object with the context at `HttpIncoming.context` which can be passed on to the client when requesting content from podlets.
```
Comment (alexanbj) [inline README.md]: incoming

File: README.md (hunk 1 above)
Comment (digitalsadhu) [reply README.md]: Method for processing an incoming HTTP...

File: README.md [hunk 3]
```diff
@@ -218,17 +225,57 @@ const layout = new Layout({
 
 The Layout instance has the following API:
 
-### .middleware()
+### .process(HttpIncoming)
 
-A connect compatible middleware which takes care of multiple operations needed for
-a Layout to fully work.
+Metod for processing a incomming http request. This method is intended to be
+used to implement support for multiple http frameworks and should not really be
```
Comment (digitalsadhu) [inline README.md]: HTTP

File: README.md [hunk 4]
```diff
@@ -218,17 +225,57 @@ const layout = new Layout({
...
+used directly in a layout server.
 
 What it does:
 
--   Runs [context parsers](https://github.com/podium-lib/context) on incomming requests and creates an object on the response at `res.locals.podium.context` which can be passed on to the client when requesting content from podlets.
--   Mounts the [proxy](https://github.com/podium-lib/proxy) so that each podlet can do transparent proxy requests if needed.
+-   Runs [@podium/context] parsers on the incomming request and sets an object with the context at `HttpIncoming.context` which can be passed on to the client when requesting content from podlets.
+-   Mounts the [@podium/proxy] so each podlet can do transparent proxy requests if needed.
+
+Returns a Promise. If the inbound request does match a proxy endpoint the
```
Comment (digitalsadhu) [inline README.md]: ...request matches a...

File: README.md [hunk 5]
```diff
@@ -310,13 +359,45 @@ const layout = new Layout({
...
-Property that exposes a metric stream. This stream joins all internal metrics streams into one stream resulting in all metrics
-from all sub modules being exposed here.
-
-Please see the [@podium/metrics module](https://github.com/podium-lib/metrics)
-for full documentation.
+Property that exposes a metric stream. This stream joins all internal metrics
+streams into one stream resulting in all metrics from all sub modules being
+exposed here.
+
+Please see [@metrics-js/metric] for full documentation.
```
Comment (digitalsadhu) [inline README.md]: @metrics/client

File: README.md [hunk 6]
```diff
@@ -310,13 +359,45 @@ const layout = new Layout({
...
+
+[express]: https://expressjs.com/ 'Express'
+[hapi layout plugin]: https://github.com/podium-lib/hapi-layout 'Hapi Layout Plugin'
+[@podium/client constructor]: https://github.com/podium-lib/client#constructor '@podium/client constructor'
+[@podium/proxy constructor]: https://github.com/podium-lib/proxy#constructor '@podium/proxy constructor'
+[@podium/context]: https://github.com/podium-lib/context '@podium/context'
+[@podium/client]: https://github.com/podium-lib/client '@podium/client'
+[@podium/proxy]: https://github.com/podium-lib/proxy '@podium/proxy'
+[httpincoming]: https://github.com/podium-lib/utils/blob/master/lib/http-incoming.js 'HttpIncoming'
+[@metrics-js/metric]: https://github.com/metrics-js/metric '@metrics-js/metric'
```
Comment (digitalsadhu) [inline README.md]: @metrics/client

File: package.json [hunk 7]
```diff
@@ -31,28 +31,29 @@
...
-    "@podium/client": "3.0.0-beta.2",
-    "@podium/context": "3.0.0-beta.1",
-    "@podium/proxy": "2.5.0",
+    "@metrics/client": "2.3.1",
+    "@podium/client": "3.0.1",
+    "@podium/context": "3.0.0",
+    "@podium/proxy": "3.0.0",
     "@podium/schemas": "3.0.0",
-    "abslog": "2.2.3",
+    "@podium/utils": "3.1.1",
```
Comment (digitalsadhu) [inline package.json]: look at all those sweet sweet clean 3.0s

**Pull request merged**